// @opensumi/monaco-editor-core/esm/vs/platform/action/common/actionCommonCategories.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Categories = void 0;
const nls_1 = require("../../../nls");
exports.Categories = Object.freeze({
    View: (0, nls_1.localize2)("vs/platform/action/common/actionCommonCategories",0, 'View'),
    Help: (0, nls_1.localize2)("vs/platform/action/common/actionCommonCategories",1, 'Help'),
    Test: (0, nls_1.localize2)("vs/platform/action/common/actionCommonCategories",2, 'Test'),
    File: (0, nls_1.localize2)("vs/platform/action/common/actionCommonCategories",3, 'File'),
    Preferences: (0, nls_1.localize2)("vs/platform/action/common/actionCommonCategories",4, 'Preferences'),
    Developer: (0, nls_1.localize2)("vs/platform/action/common/actionCommonCategories",5, "Developer"),
});
