// @opensumi/monaco-editor-core/esm/vs/editor/contrib/clipboard/browser/clipboard.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.PasteAction = exports.CopyAction = exports.CutAction = void 0;
const browser = require("../../../../base/browser/browser");
const dom_1 = require("../../../../base/browser/dom");
const platform = require("../../../../base/common/platform");
const textAreaInput_1 = require("../../../browser/controller/textAreaInput");
const editorExtensions_1 = require("../../../browser/editorExtensions");
const codeEditorService_1 = require("../../../browser/services/codeEditorService");
const editorContextKeys_1 = require("../../../common/editorContextKeys");
const copyPasteController_1 = require("../../dropOrPasteInto/browser/copyPasteController");
const nls = require("../../../../nls");
const actions_1 = require("../../../../platform/actions/common/actions");
const clipboardService_1 = require("../../../../platform/clipboard/common/clipboardService");
const contextkey_1 = require("../../../../platform/contextkey/common/contextkey");
const CLIPBOARD_CONTEXT_MENU_GROUP = '9_cutcopypaste';
const supportsCut = (platform.isNative || document.queryCommandSupported('cut'));
const supportsCopy = (platform.isNative || document.queryCommandSupported('copy'));
// Firefox only supports navigator.clipboard.readText() in browser extensions.
// See https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/readText#Browser_compatibility
// When loading over http, navigator.clipboard can be undefined. See https://github.com/microsoft/monaco-editor/issues/2313
const supportsPaste = (typeof navigator.clipboard === 'undefined' || browser.isFirefox) ? document.queryCommandSupported('paste') : true;
function registerCommand(command) {
    command.register();
    return command;
}
exports.CutAction = supportsCut ? registerCommand(new editorExtensions_1.MultiCommand({
    id: 'editor.action.clipboardCutAction',
    precondition: undefined,
    kbOpts: (
    // Do not bind cut keybindings in the browser,
    // since browsers do that for us and it avoids security prompts
    platform.isNative ? {
        primary: 2048 /* KeyMod.CtrlCmd */ | 54 /* KeyCode.KeyX */,
        win: { primary: 2048 /* KeyMod.CtrlCmd */ | 54 /* KeyCode.KeyX */, secondary: [1024 /* KeyMod.Shift */ | 20 /* KeyCode.Delete */] },
        weight: 100 /* KeybindingWeight.EditorContrib */
    } : undefined),
    menuOpts: [{
            menuId: actions_1.MenuId.MenubarEditMenu,
            group: '2_ccp',
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",0, "Cu&&t"),
            order: 1
        }, {
            menuId: actions_1.MenuId.EditorContext,
            group: CLIPBOARD_CONTEXT_MENU_GROUP,
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",1, "Cut"),
            when: editorContextKeys_1.EditorContextKeys.writable,
            order: 1,
        }, {
            menuId: actions_1.MenuId.CommandPalette,
            group: '',
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",2, "Cut"),
            order: 1
        }, {
            menuId: actions_1.MenuId.SimpleEditorContext,
            group: CLIPBOARD_CONTEXT_MENU_GROUP,
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",3, "Cut"),
            when: editorContextKeys_1.EditorContextKeys.writable,
            order: 1,
        }]
})) : undefined;
exports.CopyAction = supportsCopy ? registerCommand(new editorExtensions_1.MultiCommand({
    id: 'editor.action.clipboardCopyAction',
    precondition: undefined,
    kbOpts: (
    // Do not bind copy keybindings in the browser,
    // since browsers do that for us and it avoids security prompts
    platform.isNative ? {
        primary: 2048 /* KeyMod.CtrlCmd */ | 33 /* KeyCode.KeyC */,
        win: { primary: 2048 /* KeyMod.CtrlCmd */ | 33 /* KeyCode.KeyC */, secondary: [2048 /* KeyMod.CtrlCmd */ | 19 /* KeyCode.Insert */] },
        weight: 100 /* KeybindingWeight.EditorContrib */
    } : undefined),
    menuOpts: [{
            menuId: actions_1.MenuId.MenubarEditMenu,
            group: '2_ccp',
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",4, "&&Copy"),
            order: 2
        }, {
            menuId: actions_1.MenuId.EditorContext,
            group: CLIPBOARD_CONTEXT_MENU_GROUP,
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",5, "Copy"),
            order: 2,
        }, {
            menuId: actions_1.MenuId.CommandPalette,
            group: '',
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",6, "Copy"),
            order: 1
        }, {
            menuId: actions_1.MenuId.SimpleEditorContext,
            group: CLIPBOARD_CONTEXT_MENU_GROUP,
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",7, "Copy"),
            order: 2,
        }]
})) : undefined;
actions_1.MenuRegistry.appendMenuItem(actions_1.MenuId.MenubarEditMenu, { submenu: actions_1.MenuId.MenubarCopy, title: nls.localize2("vs/editor/contrib/clipboard/browser/clipboard",13, "Copy As"), group: '2_ccp', order: 3 });
actions_1.MenuRegistry.appendMenuItem(actions_1.MenuId.EditorContext, { submenu: actions_1.MenuId.EditorContextCopy, title: nls.localize2("vs/editor/contrib/clipboard/browser/clipboard",14, "Copy As"), group: CLIPBOARD_CONTEXT_MENU_GROUP, order: 3 });
actions_1.MenuRegistry.appendMenuItem(actions_1.MenuId.EditorContext, { submenu: actions_1.MenuId.EditorContextShare, title: nls.localize2("vs/editor/contrib/clipboard/browser/clipboard",15, "Share"), group: '11_share', order: -1, when: contextkey_1.ContextKeyExpr.and(contextkey_1.ContextKeyExpr.notEquals('resourceScheme', 'output'), editorContextKeys_1.EditorContextKeys.editorTextFocus) });
actions_1.MenuRegistry.appendMenuItem(actions_1.MenuId.EditorTitleContext, { submenu: actions_1.MenuId.EditorTitleContextShare, title: nls.localize2("vs/editor/contrib/clipboard/browser/clipboard",16, "Share"), group: '11_share', order: -1 });
actions_1.MenuRegistry.appendMenuItem(actions_1.MenuId.ExplorerContext, { submenu: actions_1.MenuId.ExplorerContextShare, title: nls.localize2("vs/editor/contrib/clipboard/browser/clipboard",17, "Share"), group: '11_share', order: -1 });
exports.PasteAction = supportsPaste ? registerCommand(new editorExtensions_1.MultiCommand({
    id: 'editor.action.clipboardPasteAction',
    precondition: undefined,
    kbOpts: (
    // Do not bind paste keybindings in the browser,
    // since browsers do that for us and it avoids security prompts
    platform.isNative ? {
        primary: 2048 /* KeyMod.CtrlCmd */ | 52 /* KeyCode.KeyV */,
        win: { primary: 2048 /* KeyMod.CtrlCmd */ | 52 /* KeyCode.KeyV */, secondary: [1024 /* KeyMod.Shift */ | 19 /* KeyCode.Insert */] },
        linux: { primary: 2048 /* KeyMod.CtrlCmd */ | 52 /* KeyCode.KeyV */, secondary: [1024 /* KeyMod.Shift */ | 19 /* KeyCode.Insert */] },
        weight: 100 /* KeybindingWeight.EditorContrib */
    } : undefined),
    menuOpts: [{
            menuId: actions_1.MenuId.MenubarEditMenu,
            group: '2_ccp',
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",8, "&&Paste"),
            order: 4
        }, {
            menuId: actions_1.MenuId.EditorContext,
            group: CLIPBOARD_CONTEXT_MENU_GROUP,
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",9, "Paste"),
            when: editorContextKeys_1.EditorContextKeys.writable,
            order: 4,
        }, {
            menuId: actions_1.MenuId.CommandPalette,
            group: '',
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",10, "Paste"),
            order: 1
        }, {
            menuId: actions_1.MenuId.SimpleEditorContext,
            group: CLIPBOARD_CONTEXT_MENU_GROUP,
            title: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",11, "Paste"),
            when: editorContextKeys_1.EditorContextKeys.writable,
            order: 4,
        }]
})) : undefined;
class ExecCommandCopyWithSyntaxHighlightingAction extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.clipboardCopyWithSyntaxHighlightingAction',
            label: nls.localize("vs/editor/contrib/clipboard/browser/clipboard",12, "Copy With Syntax Highlighting"),
            alias: 'Copy With Syntax Highlighting',
            precondition: undefined,
            kbOpts: {
                kbExpr: editorContextKeys_1.EditorContextKeys.textInputFocus,
                primary: 0,
                weight: 100 /* KeybindingWeight.EditorContrib */
            }
        });
    }
    run(accessor, editor) {
        if (!editor.hasModel()) {
            return;
        }
        const emptySelectionClipboard = editor.getOption(37 /* EditorOption.emptySelectionClipboard */);
        if (!emptySelectionClipboard && editor.getSelection().isEmpty()) {
            return;
        }
        textAreaInput_1.CopyOptions.forceCopyWithSyntaxHighlighting = true;
        editor.focus();
        editor.getContainerDomNode().ownerDocument.execCommand('copy');
        textAreaInput_1.CopyOptions.forceCopyWithSyntaxHighlighting = false;
    }
}
function registerExecCommandImpl(target, browserCommand) {
    if (!target) {
        return;
    }
    // 1. handle case when focus is in editor.
    target.addImplementation(10000, 'code-editor', (accessor, args) => {
        // Only if editor text focus (i.e. not if editor has widget focus).
        const focusedEditor = accessor.get(codeEditorService_1.ICodeEditorService).getFocusedCodeEditor();
        if (focusedEditor && focusedEditor.hasTextFocus()) {
            // Do not execute if there is no selection and empty selection clipboard is off
            const emptySelectionClipboard = focusedEditor.getOption(37 /* EditorOption.emptySelectionClipboard */);
            const selection = focusedEditor.getSelection();
            if (selection && selection.isEmpty() && !emptySelectionClipboard) {
                return true;
            }
            focusedEditor.getContainerDomNode().ownerDocument.execCommand(browserCommand);
            return true;
        }
        return false;
    });
    // 2. (default) handle case when focus is somewhere else.
    target.addImplementation(0, 'generic-dom', (accessor, args) => {
        (0, dom_1.getActiveDocument)().execCommand(browserCommand);
        return true;
    });
}
registerExecCommandImpl(exports.CutAction, 'cut');
registerExecCommandImpl(exports.CopyAction, 'copy');
if (exports.PasteAction) {
    // 1. Paste: handle case when focus is in editor.
    exports.PasteAction.addImplementation(10000, 'code-editor', (accessor, args) => {
        var _a, _b;
        const codeEditorService = accessor.get(codeEditorService_1.ICodeEditorService);
        const clipboardService = accessor.get(clipboardService_1.IClipboardService);
        // Only if editor text focus (i.e. not if editor has widget focus).
        const focusedEditor = codeEditorService.getFocusedCodeEditor();
        if (focusedEditor && focusedEditor.hasTextFocus()) {
            const result = focusedEditor.getContainerDomNode().ownerDocument.execCommand('paste');
            if (result) {
                return (_b = (_a = copyPasteController_1.CopyPasteController.get(focusedEditor)) === null || _a === void 0 ? void 0 : _a.finishedPaste()) !== null && _b !== void 0 ? _b : Promise.resolve();
            }
            else if (platform.isWeb) {
                // Use the clipboard service if document.execCommand('paste') was not successful
                return (async () => {
                    const clipboardText = await clipboardService.readText();
                    if (clipboardText !== '') {
                        const metadata = textAreaInput_1.InMemoryClipboardMetadataManager.INSTANCE.get(clipboardText);
                        let pasteOnNewLine = false;
                        let multicursorText = null;
                        let mode = null;
                        if (metadata) {
                            pasteOnNewLine = (focusedEditor.getOption(37 /* EditorOption.emptySelectionClipboard */) && !!metadata.isFromEmptySelection);
                            multicursorText = (typeof metadata.multicursorText !== 'undefined' ? metadata.multicursorText : null);
                            mode = metadata.mode;
                        }
                        focusedEditor.trigger('keyboard', "paste" /* Handler.Paste */, {
                            text: clipboardText,
                            pasteOnNewLine,
                            multicursorText,
                            mode
                        });
                    }
                })();
            }
            return true;
        }
        return false;
    });
    // 2. Paste: (default) handle case when focus is somewhere else.
    exports.PasteAction.addImplementation(0, 'generic-dom', (accessor, args) => {
        (0, dom_1.getActiveDocument)().execCommand('paste');
        return true;
    });
}
if (supportsCopy) {
    (0, editorExtensions_1.registerEditorAction)(ExecCommandCopyWithSyntaxHighlightingAction);
}
