// @opensumi/monaco-editor-core/esm/vs/editor/browser/widget/multiDiffEditorWidget/multiDiffEditorViewModel.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentDiffItemViewModel = exports.MultiDiffEditorViewModel = void 0;
const lifecycle_1 = require("../../../../base/common/lifecycle");
const observable_1 = require("../../../../base/common/observable");
const utils_1 = require("../../../../base/common/observableInternal/utils");
const diffEditorOptions_1 = require("../diffEditor/diffEditorOptions");
const diffEditorViewModel_1 = require("../diffEditor/diffEditorViewModel");
const model_1 = require("../../../common/services/model");
const instantiation_1 = require("../../../../platform/instantiation/common/instantiation");
class MultiDiffEditorViewModel extends lifecycle_1.Disposable {
    async waitForDiffs() {
        for (const d of this.items.get()) {
            await d.diffEditorViewModel.waitForDiff();
        }
    }
    collapseAll() {
        (0, observable_1.transaction)(tx => {
            for (const d of this.items.get()) {
                d.collapsed.set(true, tx);
            }
        });
    }
    expandAll() {
        (0, observable_1.transaction)(tx => {
            for (const d of this.items.get()) {
                d.collapsed.set(false, tx);
            }
        });
    }
    get contextKeys() {
        return this.model.contextKeys;
    }
    constructor(model, _instantiationService) {
        super();
        this.model = model;
        this._instantiationService = _instantiationService;
        this._documents = (0, observable_1.observableFromEvent)(this.model.onDidChange, /** @description MultiDiffEditorViewModel.documents */ () => this.model.documents);
        this.items = (0, utils_1.mapObservableArrayCached)(this, this._documents, (d, store) => store.add(this._instantiationService.createInstance(DocumentDiffItemViewModel, d)))
            .recomputeInitiallyAndOnChange(this._store);
        this.activeDiffItem = (0, observable_1.observableValue)(this, undefined);
    }
}
exports.MultiDiffEditorViewModel = MultiDiffEditorViewModel;
let DocumentDiffItemViewModel = class DocumentDiffItemViewModel extends lifecycle_1.Disposable {
    get originalUri() { var _a; return (_a = this.entry.value.original) === null || _a === void 0 ? void 0 : _a.uri; }
    get modifiedUri() { var _a; return (_a = this.entry.value.modified) === null || _a === void 0 ? void 0 : _a.uri; }
    constructor(entry, _instantiationService, _modelService) {
        var _a, _b;
        super();
        this.entry = entry;
        this._instantiationService = _instantiationService;
        this._modelService = _modelService;
        this.collapsed = (0, observable_1.observableValue)(this, false);
        this.lastTemplateData = (0, observable_1.observableValue)(this, { contentHeight: 500, selections: undefined, });
        function updateOptions(options) {
            return {
                ...options,
                hideUnchangedRegions: {
                    enabled: true,
                },
            };
        }
        const options = new diffEditorOptions_1.DiffEditorOptions(updateOptions(this.entry.value.options || {}));
        if (this.entry.value.onOptionsDidChange) {
            this._register(this.entry.value.onOptionsDidChange(() => {
                options.updateOptions(updateOptions(this.entry.value.options || {}));
            }));
        }
        const originalTextModel = (_a = this.entry.value.original) !== null && _a !== void 0 ? _a : this._register(this._modelService.createModel('', null));
        const modifiedTextModel = (_b = this.entry.value.modified) !== null && _b !== void 0 ? _b : this._register(this._modelService.createModel('', null));
        this.diffEditorViewModel = this._register(this._instantiationService.createInstance(diffEditorViewModel_1.DiffEditorViewModel, {
            original: originalTextModel,
            modified: modifiedTextModel,
        }, options));
    }
    getKey() {
        var _a, _b;
        return JSON.stringify([
            (_a = this.originalUri) === null || _a === void 0 ? void 0 : _a.toString(),
            (_b = this.modifiedUri) === null || _b === void 0 ? void 0 : _b.toString()
        ]);
    }
};
exports.DocumentDiffItemViewModel = DocumentDiffItemViewModel;
exports.DocumentDiffItemViewModel = DocumentDiffItemViewModel = __decorate([
    __param(1, instantiation_1.IInstantiationService),
    __param(2, model_1.IModelService)
], DocumentDiffItemViewModel);
