// @opensumi/monaco-editor-core/esm/vs/editor/contrib/fontZoom/browser/fontZoom.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const editorExtensions_1 = require("../../../browser/editorExtensions");
const editorZoom_1 = require("../../../common/config/editorZoom");
const nls = require("../../../../nls");
class EditorFontZoomIn extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.fontZoomIn',
            label: nls.localize("vs/editor/contrib/fontZoom/browser/fontZoom",0, "Increase Editor Font Size"),
            alias: 'Increase Editor Font Size',
            precondition: undefined
        });
    }
    run(accessor, editor) {
        editorZoom_1.EditorZoom.setZoomLevel(editorZoom_1.EditorZoom.getZoomLevel() + 1);
    }
}
class EditorFontZoomOut extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.fontZoomOut',
            label: nls.localize("vs/editor/contrib/fontZoom/browser/fontZoom",1, "Decrease Editor Font Size"),
            alias: 'Decrease Editor Font Size',
            precondition: undefined
        });
    }
    run(accessor, editor) {
        editorZoom_1.EditorZoom.setZoomLevel(editorZoom_1.EditorZoom.getZoomLevel() - 1);
    }
}
class EditorFontZoomReset extends editorExtensions_1.EditorAction {
    constructor() {
        super({
            id: 'editor.action.fontZoomReset',
            label: nls.localize("vs/editor/contrib/fontZoom/browser/fontZoom",2, "Reset Editor Font Size"),
            alias: 'Reset Editor Font Size',
            precondition: undefined
        });
    }
    run(accessor, editor) {
        editorZoom_1.EditorZoom.setZoomLevel(0);
    }
}
(0, editorExtensions_1.registerEditorAction)(EditorFontZoomIn);
(0, editorExtensions_1.registerEditorAction)(EditorFontZoomOut);
(0, editorExtensions_1.registerEditorAction)(EditorFontZoomReset);
