// @opensumi/monaco-editor-core/esm/vs/platform/quickinput/browser/quickInputUtils.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderQuickInputDescription = exports.quickInputButtonToAction = void 0;
const dom = require("../../../base/browser/dom");
const event_1 = require("../../../base/browser/event");
const event_2 = require("../../../base/common/event");
const keyboardEvent_1 = require("../../../base/browser/keyboardEvent");
const touch_1 = require("../../../base/browser/touch");
const iconLabels_1 = require("../../../base/browser/ui/iconLabel/iconLabels");
const idGenerator_1 = require("../../../base/common/idGenerator");
const linkedText_1 = require("../../../base/common/linkedText");
require("./media/quickInput.css");
const nls_1 = require("../../../nls");
const iconPathToClass = {};
const iconClassGenerator = new idGenerator_1.IdGenerator('quick-input-button-icon-');
function getIconClass(iconPath) {
    if (!iconPath) {
        return undefined;
    }
    let iconClass;
    const key = iconPath.dark.toString();
    if (iconPathToClass[key]) {
        iconClass = iconPathToClass[key];
    }
    else {
        iconClass = iconClassGenerator.nextId();
        dom.createCSSRule(`.${iconClass}, .hc-light .${iconClass}`, `background-image: ${dom.asCSSUrl(iconPath.light || iconPath.dark)}`);
        dom.createCSSRule(`.vs-dark .${iconClass}, .hc-black .${iconClass}`, `background-image: ${dom.asCSSUrl(iconPath.dark)}`);
        iconPathToClass[key] = iconClass;
    }
    return iconClass;
}
function quickInputButtonToAction(button, id, run) {
    let cssClasses = button.iconClass || getIconClass(button.iconPath);
    if (button.alwaysVisible) {
        cssClasses = cssClasses ? `${cssClasses} always-visible` : 'always-visible';
    }
    return {
        id,
        label: '',
        tooltip: button.tooltip || '',
        class: cssClasses,
        enabled: true,
        run
    };
}
exports.quickInputButtonToAction = quickInputButtonToAction;
function renderQuickInputDescription(description, container, actionHandler) {
    dom.reset(container);
    const parsed = (0, linkedText_1.parseLinkedText)(description);
    let tabIndex = 0;
    for (const node of parsed.nodes) {
        if (typeof node === 'string') {
            container.append(...(0, iconLabels_1.renderLabelWithIcons)(node));
        }
        else {
            let title = node.title;
            if (!title && node.href.startsWith('command:')) {
                title = (0, nls_1.localize)("vs/platform/quickinput/browser/quickInputUtils",0, "Click to execute command '{0}'", node.href.substring('command:'.length));
            }
            else if (!title) {
                title = node.href;
            }
            const anchor = dom.$('a', { href: node.href, title, tabIndex: tabIndex++ }, node.label);
            anchor.style.textDecoration = 'underline';
            const handleOpen = (e) => {
                if (dom.isEventLike(e)) {
                    dom.EventHelper.stop(e, true);
                }
                actionHandler.callback(node.href);
            };
            const onClick = actionHandler.disposables.add(new event_1.DomEmitter(anchor, dom.EventType.CLICK)).event;
            const onKeydown = actionHandler.disposables.add(new event_1.DomEmitter(anchor, dom.EventType.KEY_DOWN)).event;
            const onSpaceOrEnter = event_2.Event.chain(onKeydown, $ => $.filter(e => {
                const event = new keyboardEvent_1.StandardKeyboardEvent(e);
                return event.equals(10 /* KeyCode.Space */) || event.equals(3 /* KeyCode.Enter */);
            }));
            actionHandler.disposables.add(touch_1.Gesture.addTarget(anchor));
            const onTap = actionHandler.disposables.add(new event_1.DomEmitter(anchor, touch_1.EventType.Tap)).event;
            event_2.Event.any(onClick, onTap, onSpaceOrEnter)(handleOpen, null, actionHandler.disposables);
            container.appendChild(anchor);
        }
    }
}
exports.renderQuickInputDescription = renderQuickInputDescription;
