// @opensumi/monaco-editor-core/esm/vs/editor/contrib/inlayHints/browser/inlayHintsHover.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlayHintsHover = void 0;
const async_1 = require("../../../../base/common/async");
const htmlContent_1 = require("../../../../base/common/htmlContent");
const position_1 = require("../../../common/core/position");
const textModel_1 = require("../../../common/model/textModel");
const hoverTypes_1 = require("../../hover/browser/hoverTypes");
const language_1 = require("../../../common/languages/language");
const resolverService_1 = require("../../../common/services/resolverService");
const getHover_1 = require("../../hover/browser/getHover");
const markdownHoverParticipant_1 = require("../../hover/browser/markdownHoverParticipant");
const inlayHintsController_1 = require("./inlayHintsController");
const configuration_1 = require("../../../../platform/configuration/common/configuration");
const opener_1 = require("../../../../platform/opener/common/opener");
const languageFeatures_1 = require("../../../common/services/languageFeatures");
const nls_1 = require("../../../../nls");
const platform = require("../../../../base/common/platform");
const inlayHints_1 = require("./inlayHints");
const arrays_1 = require("../../../../base/common/arrays");
class InlayHintsHoverAnchor extends hoverTypes_1.HoverForeignElementAnchor {
    constructor(part, owner, initialMousePosX, initialMousePosY) {
        super(10, owner, part.item.anchor.range, initialMousePosX, initialMousePosY, true);
        this.part = part;
    }
}
let InlayHintsHover = class InlayHintsHover extends markdownHoverParticipant_1.MarkdownHoverParticipant {
    constructor(editor, languageService, openerService, configurationService, _resolverService, languageFeaturesService) {
        super(editor, languageService, openerService, configurationService, languageFeaturesService);
        this._resolverService = _resolverService;
        this.hoverOrdinal = 6;
    }
    suggestHoverAnchor(mouseEvent) {
        var _a;
        const controller = inlayHintsController_1.InlayHintsController.get(this._editor);
        if (!controller) {
            return null;
        }
        if (mouseEvent.target.type !== 6 /* MouseTargetType.CONTENT_TEXT */) {
            return null;
        }
        const options = (_a = mouseEvent.target.detail.injectedText) === null || _a === void 0 ? void 0 : _a.options;
        if (!(options instanceof textModel_1.ModelDecorationInjectedTextOptions && options.attachedData instanceof inlayHintsController_1.RenderedInlayHintLabelPart)) {
            return null;
        }
        return new InlayHintsHoverAnchor(options.attachedData, this, mouseEvent.event.posx, mouseEvent.event.posy);
    }
    computeSync() {
        return [];
    }
    computeAsync(anchor, _lineDecorations, token) {
        if (!(anchor instanceof InlayHintsHoverAnchor)) {
            return async_1.AsyncIterableObject.EMPTY;
        }
        return new async_1.AsyncIterableObject(async (executor) => {
            const { part } = anchor;
            await part.item.resolve(token);
            if (token.isCancellationRequested) {
                return;
            }
            // (1) Inlay Tooltip
            let itemTooltip;
            if (typeof part.item.hint.tooltip === 'string') {
                itemTooltip = new htmlContent_1.MarkdownString().appendText(part.item.hint.tooltip);
            }
            else if (part.item.hint.tooltip) {
                itemTooltip = part.item.hint.tooltip;
            }
            if (itemTooltip) {
                executor.emitOne(new markdownHoverParticipant_1.MarkdownHover(this, anchor.range, [itemTooltip], false, 0));
            }
            // (1.2) Inlay dbl-click gesture
            if ((0, arrays_1.isNonEmptyArray)(part.item.hint.textEdits)) {
                executor.emitOne(new markdownHoverParticipant_1.MarkdownHover(this, anchor.range, [new htmlContent_1.MarkdownString().appendText((0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",0, "Double-click to insert"))], false, 10001));
            }
            // (2) Inlay Label Part Tooltip
            let partTooltip;
            if (typeof part.part.tooltip === 'string') {
                partTooltip = new htmlContent_1.MarkdownString().appendText(part.part.tooltip);
            }
            else if (part.part.tooltip) {
                partTooltip = part.part.tooltip;
            }
            if (partTooltip) {
                executor.emitOne(new markdownHoverParticipant_1.MarkdownHover(this, anchor.range, [partTooltip], false, 1));
            }
            // (2.2) Inlay Label Part Help Hover
            if (part.part.location || part.part.command) {
                let linkHint;
                const useMetaKey = this._editor.getOption(78 /* EditorOption.multiCursorModifier */) === 'altKey';
                const kb = useMetaKey
                    ? platform.isMacintosh
                        ? (0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",1, "cmd + click")
                        : (0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",2, "ctrl + click")
                    : platform.isMacintosh
                        ? (0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",3, "option + click")
                        : (0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",4, "alt + click");
                if (part.part.location && part.part.command) {
                    linkHint = new htmlContent_1.MarkdownString().appendText((0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",5, 'Go to Definition ({0}), right click for more', kb));
                }
                else if (part.part.location) {
                    linkHint = new htmlContent_1.MarkdownString().appendText((0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",6, 'Go to Definition ({0})', kb));
                }
                else if (part.part.command) {
                    linkHint = new htmlContent_1.MarkdownString(`[${(0, nls_1.localize)("vs/editor/contrib/inlayHints/browser/inlayHintsHover",7, "Execute Command")}](${(0, inlayHints_1.asCommandLink)(part.part.command)} "${part.part.command.title}") (${kb})`, { isTrusted: true });
                }
                if (linkHint) {
                    executor.emitOne(new markdownHoverParticipant_1.MarkdownHover(this, anchor.range, [linkHint], false, 10000));
                }
            }
            // (3) Inlay Label Part Location tooltip
            const iterable = await this._resolveInlayHintLabelPartHover(part, token);
            for await (const item of iterable) {
                executor.emitOne(item);
            }
        });
    }
    async _resolveInlayHintLabelPartHover(part, token) {
        if (!part.part.location) {
            return async_1.AsyncIterableObject.EMPTY;
        }
        const { uri, range } = part.part.location;
        const ref = await this._resolverService.createModelReference(uri);
        try {
            const model = ref.object.textEditorModel;
            if (!this._languageFeaturesService.hoverProvider.has(model)) {
                return async_1.AsyncIterableObject.EMPTY;
            }
            return (0, getHover_1.getHover)(this._languageFeaturesService.hoverProvider, model, new position_1.Position(range.startLineNumber, range.startColumn), token)
                .filter(item => !(0, htmlContent_1.isEmptyMarkdownString)(item.hover.contents))
                .map(item => new markdownHoverParticipant_1.MarkdownHover(this, part.item.anchor.range, item.hover.contents, false, 2 + item.ordinal));
        }
        finally {
            ref.dispose();
        }
    }
};
exports.InlayHintsHover = InlayHintsHover;
exports.InlayHintsHover = InlayHintsHover = __decorate([
    __param(1, language_1.ILanguageService),
    __param(2, opener_1.IOpenerService),
    __param(3, configuration_1.IConfigurationService),
    __param(4, resolverService_1.ITextModelService),
    __param(5, languageFeatures_1.ILanguageFeaturesService)
], InlayHintsHover);
