// @opensumi/monaco-editor-core/esm/vs/base/common/keybindingLabels.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSettingsLabelProvider = exports.ElectronAcceleratorLabelProvider = exports.AriaLabelProvider = exports.UILabelProvider = exports.ModifierLabelProvider = void 0;
const nls = require("../../nls");
class ModifierLabelProvider {
    constructor(mac, windows, linux = windows) {
        this.modifierLabels = [null]; // index 0 will never me accessed.
        this.modifierLabels[2 /* OperatingSystem.Macintosh */] = mac;
        this.modifierLabels[1 /* OperatingSystem.Windows */] = windows;
        this.modifierLabels[3 /* OperatingSystem.Linux */] = linux;
    }
    toLabel(OS, chords, keyLabelProvider) {
        if (chords.length === 0) {
            return null;
        }
        const result = [];
        for (let i = 0, len = chords.length; i < len; i++) {
            const chord = chords[i];
            const keyLabel = keyLabelProvider(chord);
            if (keyLabel === null) {
                // this keybinding cannot be expressed...
                return null;
            }
            result[i] = _simpleAsString(chord, keyLabel, this.modifierLabels[OS]);
        }
        return result.join(' ');
    }
}
exports.ModifierLabelProvider = ModifierLabelProvider;
/**
 * A label provider that prints modifiers in a suitable format for displaying in the UI.
 */
exports.UILabelProvider = new ModifierLabelProvider({
    ctrlKey: '\u2303',
    shiftKey: '⇧',
    altKey: '⌥',
    metaKey: '⌘',
    separator: '',
}, {
    ctrlKey: nls.localize("vs/base/common/keybindingLabels",0, "Ctrl"),
    shiftKey: nls.localize("vs/base/common/keybindingLabels",1, "Shift"),
    altKey: nls.localize("vs/base/common/keybindingLabels",2, "Alt"),
    metaKey: nls.localize("vs/base/common/keybindingLabels",3, "Windows"),
    separator: '+',
}, {
    ctrlKey: nls.localize("vs/base/common/keybindingLabels",4, "Ctrl"),
    shiftKey: nls.localize("vs/base/common/keybindingLabels",5, "Shift"),
    altKey: nls.localize("vs/base/common/keybindingLabels",6, "Alt"),
    metaKey: nls.localize("vs/base/common/keybindingLabels",7, "Super"),
    separator: '+',
});
/**
 * A label provider that prints modifiers in a suitable format for ARIA.
 */
exports.AriaLabelProvider = new ModifierLabelProvider({
    ctrlKey: nls.localize("vs/base/common/keybindingLabels",8, "Control"),
    shiftKey: nls.localize("vs/base/common/keybindingLabels",9, "Shift"),
    altKey: nls.localize("vs/base/common/keybindingLabels",10, "Option"),
    metaKey: nls.localize("vs/base/common/keybindingLabels",11, "Command"),
    separator: '+',
}, {
    ctrlKey: nls.localize("vs/base/common/keybindingLabels",12, "Control"),
    shiftKey: nls.localize("vs/base/common/keybindingLabels",13, "Shift"),
    altKey: nls.localize("vs/base/common/keybindingLabels",14, "Alt"),
    metaKey: nls.localize("vs/base/common/keybindingLabels",15, "Windows"),
    separator: '+',
}, {
    ctrlKey: nls.localize("vs/base/common/keybindingLabels",16, "Control"),
    shiftKey: nls.localize("vs/base/common/keybindingLabels",17, "Shift"),
    altKey: nls.localize("vs/base/common/keybindingLabels",18, "Alt"),
    metaKey: nls.localize("vs/base/common/keybindingLabels",19, "Super"),
    separator: '+',
});
/**
 * A label provider that prints modifiers in a suitable format for Electron Accelerators.
 * See https://github.com/electron/electron/blob/master/docs/api/accelerator.md
 */
exports.ElectronAcceleratorLabelProvider = new ModifierLabelProvider({
    ctrlKey: 'Ctrl',
    shiftKey: 'Shift',
    altKey: 'Alt',
    metaKey: 'Cmd',
    separator: '+',
}, {
    ctrlKey: 'Ctrl',
    shiftKey: 'Shift',
    altKey: 'Alt',
    metaKey: 'Super',
    separator: '+',
});
/**
 * A label provider that prints modifiers in a suitable format for user settings.
 */
exports.UserSettingsLabelProvider = new ModifierLabelProvider({
    ctrlKey: 'ctrl',
    shiftKey: 'shift',
    altKey: 'alt',
    metaKey: 'cmd',
    separator: '+',
}, {
    ctrlKey: 'ctrl',
    shiftKey: 'shift',
    altKey: 'alt',
    metaKey: 'win',
    separator: '+',
}, {
    ctrlKey: 'ctrl',
    shiftKey: 'shift',
    altKey: 'alt',
    metaKey: 'meta',
    separator: '+',
});
function _simpleAsString(modifiers, key, labels) {
    if (key === null) {
        return '';
    }
    const result = [];
    // translate modifier keys: Ctrl-Shift-Alt-Meta
    if (modifiers.ctrlKey) {
        result.push(labels.ctrlKey);
    }
    if (modifiers.shiftKey) {
        result.push(labels.shiftKey);
    }
    if (modifiers.altKey) {
        result.push(labels.altKey);
    }
    if (modifiers.metaKey) {
        result.push(labels.metaKey);
    }
    // the actual key
    if (key !== '') {
        result.push(key);
    }
    return result.join(labels.separator);
}
