// @opensumi/monaco-editor-core/esm/vs/platform/quickinput/common/quickInput.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.IQuickInputService = exports.quickPickItemScorerAccessor = exports.QuickPickItemScorerAccessor = exports.ItemActivation = exports.QuickInputHideReason = exports.NO_KEY_MODS = void 0;
const instantiation_1 = require("../../instantiation/common/instantiation");
const network_1 = require("../../../base/common/network");
exports.NO_KEY_MODS = { ctrlCmd: false, alt: false };
var QuickInputHideReason;
(function (QuickInputHideReason) {
    /**
     * Focus moved away from the quick input.
     */
    QuickInputHideReason[QuickInputHideReason["Blur"] = 1] = "Blur";
    /**
     * An explicit user gesture, e.g. pressing Escape key.
     */
    QuickInputHideReason[QuickInputHideReason["Gesture"] = 2] = "Gesture";
    /**
     * Anything else.
     */
    QuickInputHideReason[QuickInputHideReason["Other"] = 3] = "Other";
})(QuickInputHideReason || (exports.QuickInputHideReason = QuickInputHideReason = {}));
/**
 * Represents the activation behavior for items in a quick input. This means which item will be
 * "active" (aka focused).
 */
var ItemActivation;
(function (ItemActivation) {
    /**
     * No item will be active.
     */
    ItemActivation[ItemActivation["NONE"] = 0] = "NONE";
    /**
     * First item will be active.
     */
    ItemActivation[ItemActivation["FIRST"] = 1] = "FIRST";
    /**
     * Second item will be active.
     */
    ItemActivation[ItemActivation["SECOND"] = 2] = "SECOND";
    /**
     * Last item will be active.
     */
    ItemActivation[ItemActivation["LAST"] = 3] = "LAST";
})(ItemActivation || (exports.ItemActivation = ItemActivation = {}));
class QuickPickItemScorerAccessor {
    constructor(options) {
        this.options = options;
    }
    getItemLabel(entry) {
        return entry.label;
    }
    getItemDescription(entry) {
        var _a;
        if ((_a = this.options) === null || _a === void 0 ? void 0 : _a.skipDescription) {
            return undefined;
        }
        return entry.description;
    }
    getItemPath(entry) {
        var _a, _b, _c;
        if ((_a = this.options) === null || _a === void 0 ? void 0 : _a.skipPath) {
            return undefined;
        }
        if (((_b = entry.resource) === null || _b === void 0 ? void 0 : _b.scheme) === network_1.Schemas.file) {
            return entry.resource.fsPath;
        }
        return (_c = entry.resource) === null || _c === void 0 ? void 0 : _c.path;
    }
}
exports.QuickPickItemScorerAccessor = QuickPickItemScorerAccessor;
exports.quickPickItemScorerAccessor = new QuickPickItemScorerAccessor();
//#endregion
exports.IQuickInputService = (0, instantiation_1.createDecorator)('quickInputService');
