// @opensumi/monaco-editor-core/esm/vs/editor/standalone/browser/standaloneLayoutService.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorScopedLayoutService = void 0;
const dom = require("../../../base/browser/dom");
const window_1 = require("../../../base/browser/window");
const arrays_1 = require("../../../base/common/arrays");
const event_1 = require("../../../base/common/event");
const codeEditorService_1 = require("../../browser/services/codeEditorService");
const extensions_1 = require("../../../platform/instantiation/common/extensions");
const layoutService_1 = require("../../../platform/layout/browser/layoutService");
let StandaloneLayoutService = class StandaloneLayoutService {
    get mainContainer() {
        var _a, _b;
        return (_b = (_a = (0, arrays_1.firstOrDefault)(this._codeEditorService.listCodeEditors())) === null || _a === void 0 ? void 0 : _a.getContainerDomNode()) !== null && _b !== void 0 ? _b : window_1.mainWindow.document.body;
    }
    get activeContainer() {
        var _a, _b;
        const activeCodeEditor = (_a = this._codeEditorService.getFocusedCodeEditor()) !== null && _a !== void 0 ? _a : this._codeEditorService.getActiveCodeEditor();
        return (_b = activeCodeEditor === null || activeCodeEditor === void 0 ? void 0 : activeCodeEditor.getContainerDomNode()) !== null && _b !== void 0 ? _b : this.mainContainer;
    }
    get mainContainerDimension() {
        return dom.getClientArea(this.mainContainer);
    }
    get activeContainerDimension() {
        return dom.getClientArea(this.activeContainer);
    }
    get containers() {
        return (0, arrays_1.coalesce)(this._codeEditorService.listCodeEditors().map(codeEditor => codeEditor.getContainerDomNode()));
    }
    getContainer() {
        return this.activeContainer;
    }
    focus() {
        var _a;
        (_a = this._codeEditorService.getFocusedCodeEditor()) === null || _a === void 0 ? void 0 : _a.focus();
    }
    constructor(_codeEditorService) {
        this._codeEditorService = _codeEditorService;
        this.onDidLayoutMainContainer = event_1.Event.None;
        this.onDidLayoutActiveContainer = event_1.Event.None;
        this.onDidLayoutContainer = event_1.Event.None;
        this.onDidChangeActiveContainer = event_1.Event.None;
        this.onDidAddContainer = event_1.Event.None;
        this.whenActiveContainerStylesLoaded = Promise.resolve();
        this.mainContainerOffset = { top: 0, quickPickTop: 0 };
        this.activeContainerOffset = { top: 0, quickPickTop: 0 };
    }
};
StandaloneLayoutService = __decorate([
    __param(0, codeEditorService_1.ICodeEditorService)
], StandaloneLayoutService);
let EditorScopedLayoutService = class EditorScopedLayoutService extends StandaloneLayoutService {
    get mainContainer() {
        return this._container;
    }
    constructor(_container, codeEditorService) {
        super(codeEditorService);
        this._container = _container;
    }
};
exports.EditorScopedLayoutService = EditorScopedLayoutService;
exports.EditorScopedLayoutService = EditorScopedLayoutService = __decorate([
    __param(1, codeEditorService_1.ICodeEditorService)
], EditorScopedLayoutService);
(0, extensions_1.registerSingleton)(layoutService_1.ILayoutService, StandaloneLayoutService, 1 /* InstantiationType.Delayed */);
