// @opensumi/monaco-editor-core/esm/vs/editor/contrib/inlineCompletions/browser/inlineCompletionsController.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var InlineCompletionsController_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineCompletionsController = void 0;
const dom_1 = require("../../../../base/browser/dom");
const aria_1 = require("../../../../base/browser/ui/aria/aria");
const lifecycle_1 = require("../../../../base/common/lifecycle");
const observable_1 = require("../../../../base/common/observable");
const coreCommands_1 = require("../../../browser/coreCommands");
const position_1 = require("../../../common/core/position");
const languageFeatureDebounce_1 = require("../../../common/services/languageFeatureDebounce");
const languageFeatures_1 = require("../../../common/services/languageFeatures");
const commandIds_1 = require("./commandIds");
const ghostTextWidget_1 = require("./ghostTextWidget");
const inlineCompletionContextKeys_1 = require("./inlineCompletionContextKeys");
const inlineCompletionsHintsWidget_1 = require("./inlineCompletionsHintsWidget");
const inlineCompletionsModel_1 = require("./inlineCompletionsModel");
const suggestWidgetInlineCompletionProvider_1 = require("./suggestWidgetInlineCompletionProvider");
const nls_1 = require("../../../../nls");
const accessibilitySignalService_1 = require("../../../../platform/accessibilitySignal/browser/accessibilitySignalService");
const commands_1 = require("../../../../platform/commands/common/commands");
const configuration_1 = require("../../../../platform/configuration/common/configuration");
const contextkey_1 = require("../../../../platform/contextkey/common/contextkey");
const instantiation_1 = require("../../../../platform/instantiation/common/instantiation");
const keybinding_1 = require("../../../../platform/keybinding/common/keybinding");
const utils_1 = require("../../../../base/common/observableInternal/utils");
let InlineCompletionsController = InlineCompletionsController_1 = class InlineCompletionsController extends lifecycle_1.Disposable {
    static get(editor) {
        return editor.getContribution(InlineCompletionsController_1.ID);
    }
    constructor(editor, _instantiationService, _contextKeyService, _configurationService, _commandService, _debounceService, _languageFeaturesService, _accessibilitySignalService, _keybindingService) {
        super();
        this.editor = editor;
        this._instantiationService = _instantiationService;
        this._contextKeyService = _contextKeyService;
        this._configurationService = _configurationService;
        this._commandService = _commandService;
        this._debounceService = _debounceService;
        this._languageFeaturesService = _languageFeaturesService;
        this._accessibilitySignalService = _accessibilitySignalService;
        this._keybindingService = _keybindingService;
        this.model = this._register((0, observable_1.disposableObservableValue)('inlineCompletionModel', undefined));
        this._textModelVersionId = (0, observable_1.observableValue)(this, -1);
        this._positions = (0, observable_1.observableValue)(this, [new position_1.Position(1, 1)]);
        this._suggestWidgetAdaptor = this._register(new suggestWidgetInlineCompletionProvider_1.SuggestWidgetAdaptor(this.editor, () => { var _a, _b; return (_b = (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.selectedInlineCompletion.get()) === null || _b === void 0 ? void 0 : _b.toSingleTextEdit(undefined); }, (tx) => this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other), (item) => {
            (0, observable_1.transaction)(tx => {
                var _a;
                /** @description InlineCompletionsController.handleSuggestAccepted */
                this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.handleSuggestAccepted(item);
            });
        }));
        this._enabled = (0, observable_1.observableFromEvent)(this.editor.onDidChangeConfiguration, () => this.editor.getOption(62 /* EditorOption.inlineSuggest */).enabled);
        this._fontFamily = (0, observable_1.observableFromEvent)(this.editor.onDidChangeConfiguration, () => this.editor.getOption(62 /* EditorOption.inlineSuggest */).fontFamily);
        this._ghostTexts = (0, observable_1.derived)(this, (reader) => {
            var _a;
            const model = this.model.read(reader);
            return (_a = model === null || model === void 0 ? void 0 : model.ghostTexts.read(reader)) !== null && _a !== void 0 ? _a : [];
        });
        this._stablizedGhostTexts = convertItemsToStableObservables(this._ghostTexts, this._store);
        this._ghostTextWidgets = (0, utils_1.mapObservableArrayCached)(this, this._stablizedGhostTexts, (ghostText, store) => {
            return store.add(this._instantiationService.createInstance(ghostTextWidget_1.GhostTextWidget, this.editor, {
                ghostText: ghostText,
                minReservedLineCount: (0, observable_1.constObservable)(0),
                targetTextModel: this.model.map(v => v === null || v === void 0 ? void 0 : v.textModel),
            }));
        }).recomputeInitiallyAndOnChange(this._store);
        this._debounceValue = this._debounceService.for(this._languageFeaturesService.inlineCompletionsProvider, 'InlineCompletionsDebounce', { min: 50, max: 50 });
        this._playAccessibilitySignal = (0, observable_1.observableSignal)(this);
        this._isReadonly = (0, observable_1.observableFromEvent)(this.editor.onDidChangeConfiguration, () => this.editor.getOption(91 /* EditorOption.readOnly */));
        this._textModel = (0, observable_1.observableFromEvent)(this.editor.onDidChangeModel, () => this.editor.getModel());
        this._textModelIfWritable = (0, observable_1.derived)(reader => this._isReadonly.read(reader) ? undefined : this._textModel.read(reader));
        this._register(new inlineCompletionContextKeys_1.InlineCompletionContextKeys(this._contextKeyService, this.model));
        this._register((0, observable_1.autorun)(reader => {
            /** @description InlineCompletionsController.update model */
            const textModel = this._textModelIfWritable.read(reader);
            (0, observable_1.transaction)(tx => {
                /** @description InlineCompletionsController.onDidChangeModel/readonly */
                this.model.set(undefined, tx);
                this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
                if (textModel) {
                    const model = _instantiationService.createInstance(inlineCompletionsModel_1.InlineCompletionsModel, textModel, this._suggestWidgetAdaptor.selectedItem, this._textModelVersionId, this._positions, this._debounceValue, (0, observable_1.observableFromEvent)(editor.onDidChangeConfiguration, () => editor.getOption(118 /* EditorOption.suggest */).preview), (0, observable_1.observableFromEvent)(editor.onDidChangeConfiguration, () => editor.getOption(118 /* EditorOption.suggest */).previewMode), (0, observable_1.observableFromEvent)(editor.onDidChangeConfiguration, () => editor.getOption(62 /* EditorOption.inlineSuggest */).mode), this._enabled);
                    this.model.set(model, tx);
                }
            });
        }));
        const styleElement = this._register((0, dom_1.createStyleSheet2)());
        this._register((0, observable_1.autorun)(reader => {
            const fontFamily = this._fontFamily.read(reader);
            styleElement.setStyle(fontFamily === '' || fontFamily === 'default' ? `` : `
.monaco-editor .ghost-text-decoration,
.monaco-editor .ghost-text-decoration-preview,
.monaco-editor .ghost-text {
	font-family: ${fontFamily};
}`);
        }));
        const getReason = (e) => {
            var _a;
            if (e.isUndoing) {
                return inlineCompletionsModel_1.VersionIdChangeReason.Undo;
            }
            if (e.isRedoing) {
                return inlineCompletionsModel_1.VersionIdChangeReason.Redo;
            }
            if ((_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.isAcceptingPartially) {
                return inlineCompletionsModel_1.VersionIdChangeReason.AcceptWord;
            }
            return inlineCompletionsModel_1.VersionIdChangeReason.Other;
        };
        this._register(editor.onDidChangeModelContent((e) => (0, observable_1.transaction)(tx => 
        /** @description InlineCompletionsController.onDidChangeModelContent */
        this.updateObservables(tx, getReason(e)))));
        this._register(editor.onDidChangeCursorPosition(e => (0, observable_1.transaction)(tx => {
            var _a;
            /** @description InlineCompletionsController.onDidChangeCursorPosition */
            this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
            if (e.reason === 3 /* CursorChangeReason.Explicit */ || e.source === 'api') {
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.stop(tx);
            }
        })));
        this._register(editor.onDidType(() => (0, observable_1.transaction)(tx => {
            var _a;
            /** @description InlineCompletionsController.onDidType */
            this.updateObservables(tx, inlineCompletionsModel_1.VersionIdChangeReason.Other);
            if (this._enabled.get()) {
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.trigger(tx);
            }
        })));
        this._register(this._commandService.onDidExecuteCommand((e) => {
            // These commands don't trigger onDidType.
            const commands = new Set([
                coreCommands_1.CoreEditingCommands.Tab.id,
                coreCommands_1.CoreEditingCommands.DeleteLeft.id,
                coreCommands_1.CoreEditingCommands.DeleteRight.id,
                commandIds_1.inlineSuggestCommitId,
                'acceptSelectedSuggestion',
            ]);
            if (commands.has(e.commandId) && editor.hasTextFocus() && this._enabled.get()) {
                (0, observable_1.transaction)(tx => {
                    var _a;
                    /** @description onDidExecuteCommand */
                    (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.trigger(tx);
                });
            }
        }));
        this._register(this.editor.onDidBlurEditorWidget(() => {
            // This is a hidden setting very useful for debugging
            if (this._contextKeyService.getContextKeyValue('accessibleViewIsShown') || this._configurationService.getValue('editor.inlineSuggest.keepOnBlur') ||
                editor.getOption(62 /* EditorOption.inlineSuggest */).keepOnBlur) {
                return;
            }
            if (inlineCompletionsHintsWidget_1.InlineSuggestionHintsContentWidget.dropDownVisible) {
                return;
            }
            (0, observable_1.transaction)(tx => {
                var _a;
                /** @description InlineCompletionsController.onDidBlurEditorWidget */
                (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.stop(tx);
            });
        }));
        this._register((0, observable_1.autorun)(reader => {
            var _a;
            /** @description InlineCompletionsController.forceRenderingAbove */
            const state = (_a = this.model.read(reader)) === null || _a === void 0 ? void 0 : _a.state.read(reader);
            if (state === null || state === void 0 ? void 0 : state.suggestItem) {
                if (state.primaryGhostText.lineCount >= 2) {
                    this._suggestWidgetAdaptor.forceRenderingAbove();
                }
            }
            else {
                this._suggestWidgetAdaptor.stopForceRenderingAbove();
            }
        }));
        this._register((0, lifecycle_1.toDisposable)(() => {
            this._suggestWidgetAdaptor.stopForceRenderingAbove();
        }));
        let lastInlineCompletionId = undefined;
        this._register((0, observable_1.autorunHandleChanges)({
            handleChange: (context, changeSummary) => {
                if (context.didChange(this._playAccessibilitySignal)) {
                    lastInlineCompletionId = undefined;
                }
                return true;
            },
        }, async (reader) => {
            /** @description InlineCompletionsController.playAccessibilitySignalAndReadSuggestion */
            this._playAccessibilitySignal.read(reader);
            const model = this.model.read(reader);
            const state = model === null || model === void 0 ? void 0 : model.state.read(reader);
            if (!model || !state || !state.inlineCompletion) {
                lastInlineCompletionId = undefined;
                return;
            }
            if (state.inlineCompletion.semanticId !== lastInlineCompletionId) {
                lastInlineCompletionId = state.inlineCompletion.semanticId;
                const lineText = model.textModel.getLineContent(state.primaryGhostText.lineNumber);
                this._accessibilitySignalService.playSignal(accessibilitySignalService_1.AccessibilitySignal.inlineSuggestion).then(() => {
                    if (this.editor.getOption(8 /* EditorOption.screenReaderAnnounceInlineSuggestion */)) {
                        this.provideScreenReaderUpdate(state.primaryGhostText.renderForScreenReader(lineText));
                    }
                });
            }
        }));
        this._register(new inlineCompletionsHintsWidget_1.InlineCompletionsHintsWidget(this.editor, this.model, this._instantiationService));
        this._register(this._configurationService.onDidChangeConfiguration(e => {
            if (e.affectsConfiguration('accessibility.verbosity.inlineCompletions')) {
                this.editor.updateOptions({ inlineCompletionsAccessibilityVerbose: this._configurationService.getValue('accessibility.verbosity.inlineCompletions') });
            }
        }));
        this.editor.updateOptions({ inlineCompletionsAccessibilityVerbose: this._configurationService.getValue('accessibility.verbosity.inlineCompletions') });
    }
    playAccessibilitySignal(tx) {
        this._playAccessibilitySignal.trigger(tx);
    }
    provideScreenReaderUpdate(content) {
        const accessibleViewShowing = this._contextKeyService.getContextKeyValue('accessibleViewIsShown');
        const accessibleViewKeybinding = this._keybindingService.lookupKeybinding('editor.action.accessibleView');
        let hint;
        if (!accessibleViewShowing && accessibleViewKeybinding && this.editor.getOption(148 /* EditorOption.inlineCompletionsAccessibilityVerbose */)) {
            hint = (0, nls_1.localize)("vs/editor/contrib/inlineCompletions/browser/inlineCompletionsController",0, "Inspect this in the accessible view ({0})", accessibleViewKeybinding.getAriaLabel());
        }
        hint ? (0, aria_1.alert)(content + ', ' + hint) : (0, aria_1.alert)(content);
    }
    /**
     * Copies over the relevant state from the text model to observables.
     * This solves all kind of eventing issues, as we make sure we always operate on the latest state,
     * regardless of who calls into us.
     */
    updateObservables(tx, changeReason) {
        var _a, _b, _c;
        const newModel = this.editor.getModel();
        this._textModelVersionId.set((_a = newModel === null || newModel === void 0 ? void 0 : newModel.getVersionId()) !== null && _a !== void 0 ? _a : -1, tx, changeReason);
        this._positions.set((_c = (_b = this.editor.getSelections()) === null || _b === void 0 ? void 0 : _b.map(selection => selection.getPosition())) !== null && _c !== void 0 ? _c : [new position_1.Position(1, 1)], tx);
    }
    shouldShowHoverAt(range) {
        var _a;
        const ghostText = (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.primaryGhostText.get();
        if (ghostText) {
            return ghostText.parts.some(p => range.containsPosition(new position_1.Position(ghostText.lineNumber, p.column)));
        }
        return false;
    }
    shouldShowHoverAtViewZone(viewZoneId) {
        var _a, _b;
        return (_b = (_a = this._ghostTextWidgets.get()[0]) === null || _a === void 0 ? void 0 : _a.ownsViewZone(viewZoneId)) !== null && _b !== void 0 ? _b : false;
    }
    hide() {
        (0, observable_1.transaction)(tx => {
            var _a;
            (_a = this.model.get()) === null || _a === void 0 ? void 0 : _a.stop(tx);
        });
    }
};
exports.InlineCompletionsController = InlineCompletionsController;
InlineCompletionsController.ID = 'editor.contrib.inlineCompletionsController';
exports.InlineCompletionsController = InlineCompletionsController = InlineCompletionsController_1 = __decorate([
    __param(1, instantiation_1.IInstantiationService),
    __param(2, contextkey_1.IContextKeyService),
    __param(3, configuration_1.IConfigurationService),
    __param(4, commands_1.ICommandService),
    __param(5, languageFeatureDebounce_1.ILanguageFeatureDebounceService),
    __param(6, languageFeatures_1.ILanguageFeaturesService),
    __param(7, accessibilitySignalService_1.IAccessibilitySignalService),
    __param(8, keybinding_1.IKeybindingService)
], InlineCompletionsController);
function convertItemsToStableObservables(items, store) {
    const result = (0, observable_1.observableValue)('result', []);
    const innerObservables = [];
    store.add((0, observable_1.autorun)(reader => {
        const itemsValue = items.read(reader);
        (0, observable_1.transaction)(tx => {
            if (itemsValue.length !== innerObservables.length) {
                innerObservables.length = itemsValue.length;
                for (let i = 0; i < innerObservables.length; i++) {
                    if (!innerObservables[i]) {
                        innerObservables[i] = (0, observable_1.observableValue)('item', itemsValue[i]);
                    }
                }
                result.set([...innerObservables], tx);
            }
            innerObservables.forEach((o, i) => o.set(itemsValue[i], tx));
        });
    }));
    return result;
}
