// @opensumi/monaco-editor-core/esm/vs/editor/browser/widget/diffEditor/registrations.contribution.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.diffDeleteDecorationEmpty = exports.diffWholeLineDeleteDecoration = exports.diffDeleteDecoration = exports.diffAddDecorationEmpty = exports.diffWholeLineAddDecoration = exports.diffAddDecoration = exports.diffLineDeleteDecorationBackground = exports.diffLineAddDecorationBackground = exports.diffLineDeleteDecorationBackgroundWithIndicator = exports.diffLineAddDecorationBackgroundWithIndicator = exports.diffRemoveIcon = exports.diffInsertIcon = exports.diffEditorUnchangedRegionShadow = exports.diffMoveBorderActive = exports.diffMoveBorder = void 0;
const codicons_1 = require("../../../../base/common/codicons");
const themables_1 = require("../../../../base/common/themables");
const textModel_1 = require("../../../common/model/textModel");
const nls_1 = require("../../../../nls");
const colorRegistry_1 = require("../../../../platform/theme/common/colorRegistry");
const iconRegistry_1 = require("../../../../platform/theme/common/iconRegistry");
exports.diffMoveBorder = (0, colorRegistry_1.registerColor)('diffEditor.move.border', { dark: '#8b8b8b9c', light: '#8b8b8b9c', hcDark: '#8b8b8b9c', hcLight: '#8b8b8b9c', }, (0, nls_1.localize)("vs/editor/browser/widget/diffEditor/registrations.contribution",0, 'The border color for text that got moved in the diff editor.'));
exports.diffMoveBorderActive = (0, colorRegistry_1.registerColor)('diffEditor.moveActive.border', { dark: '#FFA500', light: '#FFA500', hcDark: '#FFA500', hcLight: '#FFA500', }, (0, nls_1.localize)("vs/editor/browser/widget/diffEditor/registrations.contribution",1, 'The active border color for text that got moved in the diff editor.'));
exports.diffEditorUnchangedRegionShadow = (0, colorRegistry_1.registerColor)('diffEditor.unchangedRegionShadow', { dark: '#000000', light: '#737373BF', hcDark: '#000000', hcLight: '#737373BF', }, (0, nls_1.localize)("vs/editor/browser/widget/diffEditor/registrations.contribution",2, 'The color of the shadow around unchanged region widgets.'));
exports.diffInsertIcon = (0, iconRegistry_1.registerIcon)('diff-insert', codicons_1.Codicon.add, (0, nls_1.localize)("vs/editor/browser/widget/diffEditor/registrations.contribution",3, 'Line decoration for inserts in the diff editor.'));
exports.diffRemoveIcon = (0, iconRegistry_1.registerIcon)('diff-remove', codicons_1.Codicon.remove, (0, nls_1.localize)("vs/editor/browser/widget/diffEditor/registrations.contribution",4, 'Line decoration for removals in the diff editor.'));
exports.diffLineAddDecorationBackgroundWithIndicator = textModel_1.ModelDecorationOptions.register({
    className: 'line-insert',
    description: 'line-insert',
    isWholeLine: true,
    linesDecorationsClassName: 'insert-sign ' + themables_1.ThemeIcon.asClassName(exports.diffInsertIcon),
    marginClassName: 'gutter-insert',
});
exports.diffLineDeleteDecorationBackgroundWithIndicator = textModel_1.ModelDecorationOptions.register({
    className: 'line-delete',
    description: 'line-delete',
    isWholeLine: true,
    linesDecorationsClassName: 'delete-sign ' + themables_1.ThemeIcon.asClassName(exports.diffRemoveIcon),
    marginClassName: 'gutter-delete',
});
exports.diffLineAddDecorationBackground = textModel_1.ModelDecorationOptions.register({
    className: 'line-insert',
    description: 'line-insert',
    isWholeLine: true,
    marginClassName: 'gutter-insert',
});
exports.diffLineDeleteDecorationBackground = textModel_1.ModelDecorationOptions.register({
    className: 'line-delete',
    description: 'line-delete',
    isWholeLine: true,
    marginClassName: 'gutter-delete',
});
exports.diffAddDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-insert',
    description: 'char-insert',
    shouldFillLineOnLineBreak: true,
});
exports.diffWholeLineAddDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-insert',
    description: 'char-insert',
    isWholeLine: true,
});
exports.diffAddDecorationEmpty = textModel_1.ModelDecorationOptions.register({
    className: 'char-insert diff-range-empty',
    description: 'char-insert diff-range-empty',
});
exports.diffDeleteDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-delete',
    description: 'char-delete',
    shouldFillLineOnLineBreak: true,
});
exports.diffWholeLineDeleteDecoration = textModel_1.ModelDecorationOptions.register({
    className: 'char-delete',
    description: 'char-delete',
    isWholeLine: true,
});
exports.diffDeleteDecorationEmpty = textModel_1.ModelDecorationOptions.register({
    className: 'char-delete diff-range-empty',
    description: 'char-delete diff-range-empty',
});
