// @opensumi/monaco-editor-core/esm/vs/editor/browser/widget/markdownRenderer/browser/markdownRenderer.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var MarkdownRenderer_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.openLinkFromMarkdown = exports.MarkdownRenderer = void 0;
const markdownRenderer_1 = require("../../../../../base/browser/markdownRenderer");
const trustedTypes_1 = require("../../../../../base/browser/trustedTypes");
const errors_1 = require("../../../../../base/common/errors");
const event_1 = require("../../../../../base/common/event");
const lifecycle_1 = require("../../../../../base/common/lifecycle");
require("./renderedMarkdown.css");
const domFontInfo_1 = require("../../../config/domFontInfo");
const language_1 = require("../../../../common/languages/language");
const modesRegistry_1 = require("../../../../common/languages/modesRegistry");
const textToHtmlTokenizer_1 = require("../../../../common/languages/textToHtmlTokenizer");
const opener_1 = require("../../../../../platform/opener/common/opener");
/**
 * Markdown renderer that can render codeblocks with the editor mechanics. This
 * renderer should always be preferred.
 */
let MarkdownRenderer = MarkdownRenderer_1 = class MarkdownRenderer {
    constructor(_options, _languageService, _openerService) {
        this._options = _options;
        this._languageService = _languageService;
        this._openerService = _openerService;
        this._onDidRenderAsync = new event_1.Emitter();
        this.onDidRenderAsync = this._onDidRenderAsync.event;
    }
    dispose() {
        this._onDidRenderAsync.dispose();
    }
    render(markdown, options, markedOptions) {
        if (!markdown) {
            const element = document.createElement('span');
            return { element, dispose: () => { } };
        }
        const disposables = new lifecycle_1.DisposableStore();
        const rendered = disposables.add((0, markdownRenderer_1.renderMarkdown)(markdown, { ...this._getRenderOptions(markdown, disposables), ...options }, markedOptions));
        rendered.element.classList.add('rendered-markdown');
        return {
            element: rendered.element,
            dispose: () => disposables.dispose()
        };
    }
    _getRenderOptions(markdown, disposables) {
        return {
            codeBlockRenderer: async (languageAlias, value) => {
                var _a, _b, _c;
                // In markdown,
                // it is possible that we stumble upon language aliases (e.g.js instead of javascript)
                // it is possible no alias is given in which case we fall back to the current editor lang
                let languageId;
                if (languageAlias) {
                    languageId = this._languageService.getLanguageIdByLanguageName(languageAlias);
                }
                else if (this._options.editor) {
                    languageId = (_a = this._options.editor.getModel()) === null || _a === void 0 ? void 0 : _a.getLanguageId();
                }
                if (!languageId) {
                    languageId = modesRegistry_1.PLAINTEXT_LANGUAGE_ID;
                }
                const html = await (0, textToHtmlTokenizer_1.tokenizeToString)(this._languageService, value, languageId);
                const element = document.createElement('span');
                element.innerHTML = ((_c = (_b = MarkdownRenderer_1._ttpTokenizer) === null || _b === void 0 ? void 0 : _b.createHTML(html)) !== null && _c !== void 0 ? _c : html);
                // use "good" font
                if (this._options.editor) {
                    const fontInfo = this._options.editor.getOption(50 /* EditorOption.fontInfo */);
                    (0, domFontInfo_1.applyFontInfo)(element, fontInfo);
                }
                else if (this._options.codeBlockFontFamily) {
                    element.style.fontFamily = this._options.codeBlockFontFamily;
                }
                if (this._options.codeBlockFontSize !== undefined) {
                    element.style.fontSize = this._options.codeBlockFontSize;
                }
                return element;
            },
            asyncRenderCallback: () => this._onDidRenderAsync.fire(),
            actionHandler: {
                callback: (link) => openLinkFromMarkdown(this._openerService, link, markdown.isTrusted),
                disposables: disposables
            }
        };
    }
};
exports.MarkdownRenderer = MarkdownRenderer;
MarkdownRenderer._ttpTokenizer = (0, trustedTypes_1.createTrustedTypesPolicy)('tokenizeToString', {
    createHTML(html) {
        return html;
    }
});
exports.MarkdownRenderer = MarkdownRenderer = MarkdownRenderer_1 = __decorate([
    __param(1, language_1.ILanguageService),
    __param(2, opener_1.IOpenerService)
], MarkdownRenderer);
async function openLinkFromMarkdown(openerService, link, isTrusted) {
    try {
        return await openerService.open(link, {
            fromUserGesture: true,
            allowContributedOpeners: true,
            allowCommands: toAllowCommandsOption(isTrusted),
        });
    }
    catch (e) {
        (0, errors_1.onUnexpectedError)(e);
        return false;
    }
}
exports.openLinkFromMarkdown = openLinkFromMarkdown;
function toAllowCommandsOption(isTrusted) {
    if (isTrusted === true) {
        return true; // Allow all commands
    }
    if (isTrusted && Array.isArray(isTrusted.enabledCommands)) {
        return isTrusted.enabledCommands; // Allow subset of commands
    }
    return false; // Block commands
}
